From 6cc64d1d92d4fb5d055820f5cfbfed08ae555328 Mon Sep 17 00:00:00 2001
From: Dmitry Shilnenkov <dmitry.shilnenkov@gmail.com>
Date: Mon, 19 Feb 2024 21:00:21 +0200
Subject: [PATCH] Backends: Vulkan: Use sRGB color space instead of linear

---
 backends/imgui_impl_vulkan.cpp | 28 +++++++++++++++++-----------
 1 file changed, 17 insertions(+), 11 deletions(-)

diff --git a/backends/imgui_impl_vulkan.cpp b/backends/imgui_impl_vulkan.cpp
index e7f7c8d82..bb477418a 100644
--- a/backends/imgui_impl_vulkan.cpp
+++ b/backends/imgui_impl_vulkan.cpp
@@ -344,14 +344,18 @@ static uint32_t __glsl_shader_vert_spv[] =
 layout(location = 0) out vec4 fColor;
 layout(set=0, binding=0) uniform sampler2D sTexture;
 layout(location = 0) in struct { vec4 Color; vec2 UV; } In;
+
+const float GAMMA_FACTOR = 2.2;
+
 void main()
 {
-    fColor = In.Color * texture(sTexture, In.UV.st);
+    // Linear to sRGB correction
+    fColor = pow(In.Color * texture(sTexture, In.UV.st), vec4(GAMMA_FACTOR));
 }
 */
 static uint32_t __glsl_shader_frag_spv[] =
 {
-    0x07230203,0x00010000,0x00080001,0x0000001e,0x00000000,0x00020011,0x00000001,0x0006000b,
+    0x07230203,0x00010000,0x0008000a,0x00000021,0x00000000,0x00020011,0x00000001,0x0006000b,
     0x00000001,0x4c534c47,0x6474732e,0x3035342e,0x00000000,0x0003000e,0x00000000,0x00000001,
     0x0007000f,0x00000004,0x00000004,0x6e69616d,0x00000000,0x00000009,0x0000000d,0x00030010,
     0x00000004,0x00000007,0x00030003,0x00000002,0x000001c2,0x00040005,0x00000004,0x6e69616d,
@@ -369,13 +373,15 @@ static uint32_t __glsl_shader_frag_spv[] =
     0x00000007,0x00090019,0x00000013,0x00000006,0x00000001,0x00000000,0x00000000,0x00000000,
     0x00000001,0x00000000,0x0003001b,0x00000014,0x00000013,0x00040020,0x00000015,0x00000000,
     0x00000014,0x0004003b,0x00000015,0x00000016,0x00000000,0x0004002b,0x0000000e,0x00000018,
-    0x00000001,0x00040020,0x00000019,0x00000001,0x0000000a,0x00050036,0x00000002,0x00000004,
-    0x00000000,0x00000003,0x000200f8,0x00000005,0x00050041,0x00000010,0x00000011,0x0000000d,
-    0x0000000f,0x0004003d,0x00000007,0x00000012,0x00000011,0x0004003d,0x00000014,0x00000017,
-    0x00000016,0x00050041,0x00000019,0x0000001a,0x0000000d,0x00000018,0x0004003d,0x0000000a,
-    0x0000001b,0x0000001a,0x00050057,0x00000007,0x0000001c,0x00000017,0x0000001b,0x00050085,
-    0x00000007,0x0000001d,0x00000012,0x0000001c,0x0003003e,0x00000009,0x0000001d,0x000100fd,
-    0x00010038
+    0x00000001,0x00040020,0x00000019,0x00000001,0x0000000a,0x0004002b,0x00000006,0x0000001e,
+    0x400ccccd,0x0007002c,0x00000007,0x0000001f,0x0000001e,0x0000001e,0x0000001e,0x0000001e,
+    0x00050036,0x00000002,0x00000004,0x00000000,0x00000003,0x000200f8,0x00000005,0x00050041,
+    0x00000010,0x00000011,0x0000000d,0x0000000f,0x0004003d,0x00000007,0x00000012,0x00000011,
+    0x0004003d,0x00000014,0x00000017,0x00000016,0x00050041,0x00000019,0x0000001a,0x0000000d,
+    0x00000018,0x0004003d,0x0000000a,0x0000001b,0x0000001a,0x00050057,0x00000007,0x0000001c,
+    0x00000017,0x0000001b,0x00050085,0x00000007,0x0000001d,0x00000012,0x0000001c,0x0007000c,
+    0x00000007,0x00000020,0x00000001,0x0000001a,0x0000001d,0x0000001f,0x0003003e,0x00000009,
+    0x00000020,0x000100fd,0x00010038
 };
 
 //-----------------------------------------------------------------------------
@@ -694,7 +700,7 @@ bool ImGui_ImplVulkan_CreateFontsTexture()
         VkImageCreateInfo info = {};
         info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
         info.imageType = VK_IMAGE_TYPE_2D;
-        info.format = VK_FORMAT_R8G8B8A8_UNORM;
+        info.format = VK_FORMAT_R8G8B8A8_SRGB;
         info.extent.width = width;
         info.extent.height = height;
         info.extent.depth = 1;
@@ -725,7 +731,7 @@ bool ImGui_ImplVulkan_CreateFontsTexture()
         info.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
         info.image = bd->FontImage;
         info.viewType = VK_IMAGE_VIEW_TYPE_2D;
-        info.format = VK_FORMAT_R8G8B8A8_UNORM;
+        info.format = VK_FORMAT_R8G8B8A8_SRGB;
         info.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
         info.subresourceRange.levelCount = 1;
         info.subresourceRange.layerCount = 1;
-- 
2.39.3 (Apple Git-145)

